Changing Access Modifiers

cs17resch01003 opened the pull request:
As suggested in mc-imperial/gpuverify#38, we have broken the PR into two parts. This pull request includes the changes to the access modifiers of the GPUVerifyErrorReporter class and some methods in it so that it can be extended in our (@cs17resch01003 and @sbjoshi) work on GPURepair (https://arxiv.org/abs/2011.08373), We have also added a public method in the SourceLocationInfo class to expose the records that make up the source location.

We have tried to maintain the same code formatting as that of the existing code. Please let us know if any changes are needed or if there are any questions.

One thing we noticed is that the Travis build configuration checks out the code for libclc from version bac1578088e of https://git.llvm.org/git/libclc.git/ which doesn't exist anymore. For GPURepair, we are using https://github.com/llvm/llvm-project/tree/release/8.x/libclc which we believe is the closest to the checked-out version. With that change one of the test cases (OpenCL/pointeranalysistests/manyprocedures) is failing. GPUVerify is able to identify null pointer accesses and write-write races for this kernel. This test case is failing even without our code changes. Attaching the [output ](https://github.com/mc-imperial/gpuverify/files/7278836/manyprocedures.log) from GPUVerify. Please let us know if any further investigation is needed from our side.

jeroenk commented:
A question for some context: I assume you want access to this information, because it's the actual information based on which you do the repair?

cs17resch01003 commented:
@jeroenk that's correct. Instead of relying on the text output of GPUVerify, we are extending this class to get the Boogie model and the source location of the errors.

jeroenk commented:
Thanks. I feel a bit uncomfortable with the guts of `GPUVerifyErrorReporter` being exposed, and it's also not very nice for you to have to duplicate some of our code. I'll have a little think to see if I can provide you with a better interface.

jeroenk commented on GPUVerifyLib/GPUVerifyErrorReporter.cs:
```diff
@@ -527,12 +527,12 @@ private static string GetStateName(AssertCounterexample assertCex)
         }
 
-        private static string GetSourceFileName()
+        protected static string GetSourceFileName()
```
Exposing this seems a bit dubious. It assumes that the way we handle the command line is identical to the way you handle it. It might be better to define a `GetSourceFileName` method on yourself and that you are in complete control over.

jeroenk commented on GPUVerifyLib/GPUVerifyErrorReporter.cs:
```diff
@@ -802,7 +802,7 @@ private static AccessType GetAccessType(CallCounterexample err)
         }
 
-        private static QKeyValue GetAttributes(Absy a)
+        protected static QKeyValue GetAttributes(Absy a)
```
Is it necessary to expose this? In the two locations where you use this, can't you just write `example.FailingCall.Attributes`?

cs17resch01003 replied:
Reverting this change.

jeroenk commented on GPUVerifyLib/GPUVerifyErrorReporter.cs:
```diff
@@ -541,7 +541,7 @@ private static void PopulateModelWithStatesIfNecessary(Counterexample cex)
         }
 
-        private static void DetermineNatureOfRace(CallCounterexample callCex, out string raceName, out string access1, out string access2)
+        protected static void DetermineNatureOfRace(CallCounterexample callCex, out string raceName, out string access1, out string access2)
```
Instead of calling this method and then `AccessType.Create(access1)`, it would be better to expose `GetAccessType` and call `GetAccessType(callCex)`. Although you currently store the result of `DetermineNatureOfRace` on you side, you don't seem to do anything with the result, so it should be easy to remove.

cs17resch01003 replied:
Leaving this method as private and making GetAccessType as protected.

cs17resch01003 commented:
@jeroenk this pull request got automatically closed when I recreated the branch to avoid merge conflicts. I addressed all the comments in this PR and raised a new one at https://github.com/mc-imperial/gpuverify/pull/49.
